library/src/cache/patch_manager.rs

```diff
@@ -0,0 +1,683 @@
+use super::{disk_io, PatchInfo};
+use anyhow::{bail, Context, Ok, Result};
+use core::fmt::Debug;
+use serde::{Deserialize, Serialize};
+use std::path::{Path, PathBuf};
+
+#[cfg(test)]
+use mockall::automock;
+#[cfg(test)]
+use tempdir::TempDir;
+
+// https://stackoverflow.com/questions/67087597/is-it-possible-to-use-rusts-log-info-for-tests
+#[cfg(test)]
+use std::{println as info, println as error}; // Workaround to use println! for logs.
+
+const PATCHES_DIR_NAME: &str = "patches";
+const PATCHES_STATE_FILE_NAME: &str = "patches_state.json";
+
+#[derive(Copy, Clone, Debug, Deserialize, Serialize, PartialEq, Eq, Hash)]
+struct PatchMetadata {
+    /// The number of the patch.
+    number: usize,
+
+    /// The size of the patch artifact on disk.
+    size: u64,
+}
+
+/// What gets serialized to disk
+#[derive(Debug, Default, Deserialize, Serialize)]
+struct PatchesState {
+    /// The patch we are currently running, if any.
+    last_booted_patch: Option<PatchMetadata>,
+
+    /// The patch that will be run on the next app boot, if any. This may be the same
+    /// as the last booted patch patch if no new patch has been downloaded.
+    next_boot_patch: Option<PatchMetadata>,
+
+    /// The highest patch number we have seen. This may be higher than the last booted
+    /// patch or next patch if we downloaded a patch that failed to boot.
+    highest_seen_patch_number: Option<usize>,
+}
+
+/// Abstracts the process of managing patches.
+#[cfg_attr(test, automock)]
+pub trait ManagePatches {
+    /// Copies the patch file at file_path to the manager's directory structure sets
+    /// this patch as the next patch to boot.
+    fn add_patch(&mut self, number: usize, file_path: &Path) -> Result<()>;
+
+    /// Returns the patch we most recently successfully booted from (usually the currently running patch),
+    /// or None if no patch is installed.
+    fn last_successfully_booted_patch(&self) -> Option<PatchInfo>;
+
+    /// Returns the next patch to boot, or None if no new patch has been downloaded.
+    fn get_next_boot_patch(&mut self) -> Option<PatchInfo>;
```

Comment:
I don't know rust style.  Does rust use get_ in names?

Reply:
It generally does not. Removed.